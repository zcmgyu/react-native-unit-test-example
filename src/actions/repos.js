// Constants
export const SELECT_REPO = 'SELECT_REPO';

// Actions
export const selectRepo = id => {
  return {
    type: SELECT_REPO,
    id
  };
};
