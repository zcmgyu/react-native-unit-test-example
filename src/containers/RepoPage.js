import { connect } from 'react-redux';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { StyleSheet, Text, View } from 'react-native';

import { fetchData, receiveRepos } from '../actions/api';
import { selectRepo } from '../actions/repos';
import RepoList from '../components/RepoList';
import { repositories } from '../utils/endpoints';

// Exporting for testing purposes
export class RepoPage extends Component {
  componentDidMount() {
    this.props.getRepos();
  }

  render() {
    const { isLoading, repos, selected } = this.props.repos;
    return (
      <View style={styles.container}>
        {!repos || isLoading ? (
          <Text style={styles.loading}>Loading...</Text>
        ) : (
          <RepoList
            repos={repos}
            selected={selected}
            selectRepo={this.props.selectRepo}
          />
        )}
      </View>
    );
  }
}

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center'
  },
  loading: {
    fontSize: 20,
    alignSelf: 'center'
  }
});

RepoPage.propTypes = {
  getRepos: PropTypes.func.isRequired,
  repos: PropTypes.object.isRequired,
  selectRepo: PropTypes.func,
  selected: PropTypes.bool
}

const mapStateToProps = state => ({ repos: state.repos });
const mapDispatchToProps = dispatch => ({
  getRepos: () => dispatch(fetchData(repositories, receiveRepos)),
  selectRepo: id => dispatch(selectRepo(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(RepoPage);
