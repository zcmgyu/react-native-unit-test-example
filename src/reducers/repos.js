import { LOAD_REPOS } from '../actions/api';
import { SELECT_REPO } from '../actions/repos';

export const initialState = {
  // Exporting it for test purposes
  error: null,
  isLoading: false,
  repos: null,
  selected: null
};

export default function repos(state = initialState, action) {
  switch (action.type) {
    case LOAD_REPOS.failure:
      return { ...state, error: action.error, isLoading: false };
    case LOAD_REPOS.request:
      return { ...state, isLoading: true };
    case LOAD_REPOS.success:
      return { ...state, isLoading: false, repos: action.data };
    case SELECT_REPO:
      const selected = state.selected !== action.id ? action.id : null;
      return { ...state, selected };
    default:
      return state;
  }
}
