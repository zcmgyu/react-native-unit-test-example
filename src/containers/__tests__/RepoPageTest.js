import React from 'react';
import { Text } from 'react-native';
import { RepoPage } from '../RepoPage';
import repos from '../../../config/jest/dataMock';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

jest.mock('../../components/RepoList', () => 'RepoList');

it('renders a RepoPage using Snapshots', () => {
  const _getComponent = props => {
    return renderer.create(<RepoPage repos={props} getRepos={jest.fn()} />);
  };

  let component = _getComponent({ isLoading: true, repos: [] });
  expect(component).toMatchSnapshot();

  component = _getComponent({ isLoading: false, repos });
  expect(component).toMatchSnapshot();
});

it('Test func inside RepoPage', () => {
  const props = { isLoading: false, repos: [] };
  const wrapper = renderer.create(
    <RepoPage repos={props} getRepos={jest.fn()} />
  );
  const instance = wrapper.getInstance();
  expect(instance.handleSomething('John Doe')).toMatchSnapshot();
});
