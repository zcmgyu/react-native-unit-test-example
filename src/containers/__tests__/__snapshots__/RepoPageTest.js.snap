// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a RepoPage using Snapshots 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "alignSelf": "center",
        "fontSize": 20,
      }
    }
  >
    Loading...
  </Text>
</View>
`;

exports[`renders a RepoPage using Snapshots 2`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <RepoList
    repos={
      Array [
        Object {
          "description": "Useful description goes here.",
          "id": 1,
          "name": "Test 1",
          "stargazers_count": 50000,
        },
        Object {
          "description": "Useful description goes here.",
          "id": 2,
          "name": "Test 2",
          "stargazers_count": 10000,
        },
        Object {
          "description": "Useful description goes here.",
          "id": 3,
          "name": "Test 3",
          "stargazers_count": 499,
        },
      ]
    }
    selectRepo={undefined}
    selected={undefined}
  />
</View>
`;
